### PR TITLE
Add `vue/no-restricted-component-names` rule

### LINKED ISSUE #2182: Rule Proposal: `vue/no-restricted-component-names`
**Please describe what the rule should do:**

Restrict some component names via a list of pattern.

**What category should the rule belong to?**

[ ] Enforces code style (layout)
[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule should warn about:**

Options:

```json
{
  "vue/no-restricted-component-names": ["error", "Foo", "/^Baz(One|Two)/"]
}
```

```vue
<script>
<!-- ✗ BAD -->
export default {
  name: 'Foo'
}

<!-- ✗ BAD -->
export default {
  name: 'BazTwo'
}

<!-- ✓ GOOD-->
export default {
  name: 'Allow'
}
</script>
```

**Additional context**

Related https://github.com/vuejs/eslint-plugin-vue/issues/2180#issuecomment-1551298495.

### PR DESCRIPTION
**Pull request by ItMaga**
closes #2182 

---

I slightly changed the schema format that was proposed in the issue, as in [no-restricted-props](https://github.com/vuejs/eslint-plugin-vue/blob/master/lib/rules/no-restricted-props.js) rule.

### CODE REVIEW

File: lib/rules/no-restricted-component-names.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
+/**
+ * @author ItMaga <https://github.com/ItMaga>
+ * See LICENSE file in root directory for full license.
+ */
+'use strict'
+
+const utils = require('../utils')
+const { isRegExp } = require('../utils/regexp')
+const { toRegExp } = require('../utils/regexp')
```
Comment (FloEdelmann) [inline lib/rules/no-restricted-component-names.js:9]:
```suggestion
const { isRegExp, toRegExp } = require('../utils/regexp')
```

File: lib/rules/no-restricted-component-names.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+/**
+ * @param {Property | AssignmentProperty} property
+ * @param {string | undefined} suggest
+ * @returns {Rule.SuggestionReportDescriptor[]}
+ * @private
+ * */
+function createSuggest(property, suggest) {
+  if (!suggest) return []
```
Comment (FloEdelmann) [inline lib/rules/no-restricted-component-names.js:54]:
```suggestion
  if (!suggest) {
    return []
  }
```

File: lib/rules/no-restricted-component-names.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    const options = context.options.map(parseOption)
+
+    /**
+     * @param {Property | AssignmentProperty} property
+     */
+    function verify(property) {
+      const propertyName = utils.getStaticPropertyName(property)
+      if (propertyName === 'name' && property.value.type === 'Literal') {
+        const componentName = property.value.value?.toString()
+        if (!componentName) return
```
Comment (FloEdelmann) [inline lib/rules/no-restricted-component-names.js:115]:
```suggestion
        if (!componentName) {
          return
        }
```

File: lib/rules/no-restricted-component-names.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    /** @type {OptionParsed[]} */
+    const options = context.options.map(parseOption)
+
+    /**
+     * @param {Property | AssignmentProperty} property
+     */
+    function verify(property) {
+      const propertyName = utils.getStaticPropertyName(property)
+      if (propertyName === 'name' && property.value.type === 'Literal') {
+        const componentName = property.value.value?.toString()
```
Comment (FloEdelmann) [inline lib/rules/no-restricted-component-names.js:114]: I think that conversion between kebab-case and PascalCase should also be supported, i.e. `DisallowedComponent` should also report for `disallowed-component` and vice-versa. Please also add tests for that.

**Review by FloEdelmann — CHANGES_REQUESTED** [review]
Thanks for the contribution! I have some comments, see below.

File: lib/rules/no-restricted-component-names.js (hunk 4 above)
Comment (ItMaga) [reply lib/rules/no-restricted-component-names.js:116]: Thank you! Resolved

**Review by FloEdelmann — APPROVED** [review]
Looks good to me, thanks! :slightly_smiling_face:

File: lib/rules/no-restricted-component-names.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+              suggest: createSuggest(property, option.suggest)
+            })
+          }
+        }
+      }
+    }
+
+    return utils.compositingVisitors(
+      utils.defineVueVisitor(context, {
+        Property(node) {
```
Comment (ota-meshi) [inline lib/rules/no-restricted-component-names.js:140]: Could you change it to use the `onVueObjectEnter` listener and only check the properties of the Vue object?
The following code may give false positives.

```vue
<script>
export default {
  data () {
    return {
      name: 'Disallow'
    }
  }
}
</script>
```

**Review by ota-meshi — CHANGES_REQUESTED** [review]

File: lib/rules/no-restricted-component-names.js (hunk 5 above)
Comment (ItMaga) [reply lib/rules/no-restricted-component-names.js]: Thank you! Done.

**Review by ota-meshi — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**